bokeh/core/has_props.py

```diff
@@ -607,8 +607,17 @@ def apply_theme(self, property_values):
 
         # Emit any change notifications that result
         for k, v in old_values.items():
-            prop = self.lookup(k)
-            prop.trigger_if_changed(self, v)
+            descriptor = self.lookup(k)
+            if isinstance(v, PropertyValueContainer) and v._unmodified_default_value:
+                if k in self._property_values[k]:
+                    self._property_values[k]._unregister_owner(self, descriptor)
+                if k in property_values:
+                    self._property_values[k] = descriptor.property.prepare_value(self.__class__, k, property_values[k])
+                else:
+                    self._property_values[k] = descriptor.class_default(self.__class__)
+                self._property_values[k]._unmodified_default_value = True
+                self._property_values[k]._register_owner(self, descriptor)
+            descriptor.trigger_if_changed(self, v)
```

Comment:
what was the actual diagnosis for the cause of #5644 ? what goes wrong?

One test case I'd recommend is to apply the same theme dict instance to multiple models (as suggested in the doc comment for apply_theme here), and be sure nothing goes haywire. I guess we are trying to avoid copying the theme once per model but this does mean the theme dict can't be modified.